### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by pcd1193182**
This PR contains the initial implementation of the Linux Recovery Environment project. Testing has been performed by building the package, and loading it onto a Delphix VM with a modified version of GRUB and ZFS to support the bootenv feature. Also tested with ab-pre-push: http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-orchestrator-pre-push/3621/console

### CODE REVIEW

File: scripts/stage1.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+target="$1"
+workdir=$(mktemp -d)
+img=$workdir/img
+base="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+#
+# Set up working directory
+#
+cd $workdir
+sudo mkdir img
```
Comment (prakashsurya) [inline scripts/stage1.sh:21]: Is `sudo` required here and in the calls below? If so, can you comment on why we need it.. if not, can you remove them..

IMO, we should (ideally) be able to build the package (based on my understanding, this is only called during the package build, is that correct?) without needing root privileges.. if we do need root, I'd like better understand why it's required.

File: scripts/stage1.sh [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+sudo rsync -aL /lib64/ld-linux-x86-64.so.2 $img/lib64/
+sudo mkdir -p $img/lib/x86_64-linux-gnu/
+ls $img/lib/x86_64-linux-gnu/
+sudo rsync -aL /lib/x86_64-linux-gnu/libnss* $img/lib/x86_64-linux-gnu/
+ls $img/lib/x86_64-linux-gnu/
+sudo rsync -a /lib/modprobe.d $img/lib/
+
+sudo rsync -a /etc/{group,passwd,shadow}{,-} $img/etc/
+sudo sed -i 's@/bin/bash@/bin/sh@g' $img/etc/passwd{,-}
+sudo sed -i 's@/opt/delphix/server/bin/supportlogin@/bin/sh@g' $img/etc/passwd{,-}
```
Comment (prakashsurya) [inline scripts/stage1.sh:50]: This puts a dependency on the virtualization package, right?

Additionally, is this script run during the package build phase? If so, this will use the file from the build server, right (I see it called only from the Makefile)? Is that correct? I'm not sure I fully understand where/when this script will run...

File: base_img/init [hunk 3]
```diff
@@ -0,0 +1,36 @@
+#!/bin/sh
```
Comment (prakashsurya) [inline base_img/init:1]: There's a mix of `/bin/sh` and `/bin/bash` used in the scripts in this PR.. is that intentional? can we pick one or the other, just for consistency sake?

File: base_img/init_functions [hunk 4]
```diff
@@ -0,0 +1,418 @@
+# This file contains common functions used in init and in hooks
```
Comment (prakashsurya) [inline base_img/init_functions:1]: Did you create all the code in this file, or was it copied from somewhere else?

File: debian/postinst [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+{
+    echo "$1" >&2
+    exit 1
+}
+
+target="/boot/recovery.img"
+workdir=$(mktemp -d)
+
+cd $workdir
+sudo chown root .
```
Comment (prakashsurya) [inline debian/postinst:13]: Similar to another comment I left, do we need to use `sudo` here? IIRC, we'll run this as root via the post-install package hook, so I think the use of `sudo` can be removed (and should be removed, if it's unnecessary, IMO).

File: debian/rules [hunk 6]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/make -f
+
+%:
+	dh $@
+
```
Comment (prakashsurya) [inline debian/rules:5]: There's some trailing newlines in various files, including this one.. can we remove them?

File: scripts/get_deps [hunk 7]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+set -euxo pipefail
+
+function die
```
Comment (prakashsurya) [inline scripts/get_deps:5]: is this function used?

File: scripts/recovery_sync [hunk 8]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+
+set -euxo pipefail
+
+function die
```
Comment (prakashsurya) [inline scripts/recovery_sync:5]: is this function used?

File: scripts/recovery_sync [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+workdir=$(mktemp -d)
+
+cd $workdir
+sudo chown root .
+sudo chmod 0755 .
+cat $target | gunzip | sudo cpio -idmv
+
+sudo rsync -a /run/systemd/network/ etc/systemd/network
+sudo mkdir -p etc/dropbear
+for fmt in $(ls /etc/ssh/*key | sed 's/.*host_\(.*\)_key/\1/'); do
```
Comment (prakashsurya) [inline scripts/recovery_sync:22]: Are we copying configuration from the build server here?

File: scripts/stage1.sh (hunk 2 above)
Comment (pcd1193182) [reply scripts/stage1.sh:50]: This code should be in the postinst script, I've also added a comment that clarifies which script is run where.

File: base_img/init (hunk 3 above)
Comment (pcd1193182) [reply base_img/init:1]: Switched to `/bin/bash` in all cases except for the base_img scripts, which are running in the busybox context, and so don't have access to bash.

File: base_img/init_functions (hunk 4 above)
Comment (pcd1193182) [reply base_img/init_functions:1]: This code (along with almost everything else in base_img) was copied from mkinitcpio.

File: scripts/get_deps (hunk 7 above)
Comment (pcd1193182) [reply scripts/get_deps:5]: Yes, immediately below this if bad parameters are passed.

Comment (prakashsurya) [reply scripts/get_deps:5]: I think I meant to leave this comment in the `postinst` script.. the `die` function isn't used in that script, right?

File: scripts/recovery_sync (hunk 8 above)
Comment (pcd1193182) [reply scripts/recovery_sync:5]: Same as in get_deps

File: scripts/recovery_sync (hunk 9 above)
Comment (pcd1193182) [reply scripts/recovery_sync:22]: No, recovery_sync is run on the engine when the networking or other config is changed.

Comment (prakashsurya) [reply scripts/recovery_sync:22]: Ah, ok.. I must not have looked closely enough.. I see now the Makefile is just copying it into place, rather than executing it. Sorry for the confusion.

**Comment by pcd1193182** [comment]
> I think this looks reasonable to me. It took a little bit of time for me to come up to speed; I think some explanation of the repository layout and a high level overview of the files/directories would have helped, but I think I understand better now.

I'll add something to the README explaining the overall structure, I agree it's not at all obvious just from the structure or naming of the code what's going on.

File: scripts/stage1.sh [hunk 10]
```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
+#
+# This script is run at package build time to create the base recovery
```
[the remark this answers is not in the document]
Comment (pcd1193182) [reply scripts/stage1.sh:3]: My understanding was that the bootstrap VM was constant for a given LTS (for the relevant files; we might get new packages, but the kernel version would be staying the same). If that is not the case, then we need to come up with a different design if we want repeatable builds.

File: debian/postinst [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+rsync -a /$modpath/modules* ./$modpath/
+rsync -a /$modpath/vdso ./$modpath/
+cp -r /$modpath/kernel/drivers/{char,misc,tty,net} ./$modpath/kernel/drivers/
+
+if [[ $(uname -r) =~ .*gcp ]]
+then
+    rsync -a /$modpath/kernel/net/ ./$modpath/kernel/net/
+fi
+
+rm -rf ./$modpath/kernel/drivers/net/wireless #TODO expand?
```
[the remark this answers is not in the document]
Comment (pcd1193182) [reply debian/postinst:30]: This was a note to myself to figure out which parts of the networking drivers we could safely remove from the recovery environment without causing problems. I'll remove it.

File: scripts/stage1.sh (hunk 10 above)
[the remark this answers is not in the document]
Comment (pzakha) [reply scripts/stage1.sh:3]: @pcd1193182 I'd not rely on files from the build server since we have rebuilt it a few times before and may configure it in such a way that is different from what we'd want in the recovery environment.

For configuration files copied from `/etc/` we could perhaps keep static copies in this repository. This would also allow us to customize them as needed.

For things like libraries we could perhaps `apt-get download`/extract the packages and then copy them from there? This way we'd use whatever is provided by the linux-package-mirror like we do for the rest of our appliance dependencies.

File: base_img/hooks/ssh [hunk 12]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+run_hook() {
+    mkdir /dev/pts
+    mount -t devpts /dev/pts /dev/pts
+    dropbear -E
+    sshd_running=1
```
Comment (pzakha) [inline base_img/hooks/ssh:7]: what does that line do?

File: base_img/hooks/networkd [hunk 13]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/sh
+
+run_hook() {
+    mkdir -p /run/systemd
+    /lib/systemd/systemd-networkd &
+    networkd_running=1
```
Comment (pzakha) [inline base_img/hooks/networkd:6]: what does that line do?

File: base_img/config [hunk 14]
```diff
@@ -0,0 +1,4 @@
+EARLYHOOKS="udev"
+HOOKS="udev networkd ssh"
```
Comment (pzakha) [inline base_img/config:2]: It seems like the order of the hooks matter here. Could you add a comment?

File: debian/postinst [hunk 15]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
```
Comment (pzakha) [inline debian/postinst:1]: I believe this would be re-run on every upgrade if the recovery-environment package's version is bumped. I think what we wanted is to have this run exclusively during the initial image creation, right?

File: debian/control [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+Priority: optional
+Maintainer: Paul Dagnelie <pcd@delphix.com>
+Build-Depends: debhelper (>= 10), dropbear-bin
+Standards-Version: 4.1.2
+Vcs-Git: https://github.com/delphix/recovery-environment.git
+#Vcs-Browser: https://github.com/delphix/recovery-environment
+
+Package: recovery-environment
+Architecture: any
+Depends: ${shlibs:Depends}, ${misc:Depends}
```
Comment (pzakha) [inline debian/control:12]: This package copies a bunch of files from other packages in the postinst script (such as from the kernel). Therefore we need to make sure that it is installed after those other packages are installed, so we'd want to put those dependencies here. For instance, we would want to add `delphix-kernel` as a dependency here, as well as other packages that we are copying files from.

File: debian/copyright [hunk 17]
```diff
@@ -0,0 +1,47 @@
+Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
+Upstream-Name: recovery-environment
+Source: <url://example.com>
+
+Files: *
+Copyright: 2020 Delphix 
+License: Apache2
```
Comment (pzakha) [inline debian/copyright:7]: See my comment in the LICENSE file regarding the license.

File: debian/recovery-environment.substvars [hunk 18]
```diff
@@ -0,0 +1,2 @@
+misc:Depends=
+misc:Pre-Depends=
```
Comment (pzakha) [inline debian/recovery-environment.substvars:2]: Why is that needed?

File: scripts/get_deps [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+    echo "$1" >&2
+    exit 1
+}
+
+[[ $# -eq 2 ]] || die "Illegal number of parameters"
+[[ -f $1 ]] || die "No such file: $1"
+
+(ldd $1 2>/dev/null || true) | while read line;
+do
+    if ! echo "$line" | grep "=>" >/dev/null 2>/dev/null ; then
```
Comment (pzakha) [inline scripts/get_deps:16]: trick: can use `&>/dev/null` instead of `>/dev/null 2>/dev/null`

File: scripts/get_deps [hunk 20]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
```
Comment (pzakha) [inline scripts/get_deps:1]: Could this script be a function inside `scripts/stage1` instead?

File: scripts/recovery_sync [hunk 21]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
```
Comment (pzakha) [inline scripts/recovery_sync:1]: Could you please add a short description for this script?

File: Makefile [hunk 22]
```diff
@@ -0,0 +1,36 @@
+prefix = /usr/
```
[the remark this answers is not in the document]
Comment (pcd1193182) [reply Makefile:1]: It is somewhat traditional in makefiles to have a prefix variable so the installation directory can be changed easily by the distro/package builders if they desire. I don't think we would ever reconfigure it (except maybe for testing purposes), but others might.

File: LICENSE [hunk 23]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
[the remark this answers is not in the document]
Comment (pcd1193182) [reply LICENSE:1]: That is probably the easiest and safest approach, yeah. I am fuzzy on exactly what constitutes a derived work, since I am not a lawyer.

File: base_img/hooks/ssh (hunk 12 above)
Comment (pcd1193182) [reply base_img/hooks/ssh:7]: In general, the hook functions define this variable so that other init-related functions can readily check if the service has been launched without having to resort to something like pgrep. We don't specifically use this variable, currently, but it is defined for consistency and for potential future use.

File: base_img/hooks/networkd (hunk 13 above)
Comment (pcd1193182) [reply base_img/hooks/networkd:6]: See previous

File: debian/postinst (hunk 15 above)
Comment (pcd1193182) [reply debian/postinst:1]: We want to re-copy over major files on upgrade, so that the recovery environment doesn't go too far out of date with the rest of the system.

File: scripts/stage1.sh [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+# shellcheck disable=SC2001
+
+set -euxo pipefail
+
+function die() {
+	echo "$1" >&2
+	exit 1
+}
+
+function get_deps() {
```
Comment (pzakha) [inline scripts/stage1.sh:17]: nit: can you mention what are the arguments, either via comments or via self-explanatory local variables like
```
local binary="$1"
local destdir="$2
```

File: scripts/stage1.sh [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+	(ldd "$1" 2>/dev/null || true) | while read -r line; do
+		if ! echo "$line" | grep "=>" &>/dev/null; then
+			continue
+		fi
+		dep=$(echo "$line" | sed 's/.*=> \(.*\) (.*/\1/')
+		dn=$(dirname "$dep")
```
Comment (pzakha) [inline scripts/stage1.sh:23]: nit: use `local` for local variables `dep`, `dn`, `line`

File: scripts/stage1.sh [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+		mkdir -p "$2/$dn"
+		rsync -aL "$dep" "$2/$dep"
```
Comment (pzakha) [inline scripts/stage1.sh:25]: Is this going to rsync from the build system or from the `workdir`?

File: scripts/stage1.sh [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+rsync -aL "$workdir/lib64/ld-linux-x86-64.so.2" "$img/lib64/"
+mkdir -p "$img/lib/x86_64-linux-gnu/"
+rsync -aL "$workdir"/lib/x86_64-linux-gnu/libnss* "$img/lib/x86_64-linux-gnu/"
+rsync -a "$workdir/lib/modprobe.d" "$img/lib/"
+
+mkdir -p "$img"/etc/{systemd/network,network,dhcp}
+mkdir -p "$img/etc/dropbear"
+
+cd "$img"
+find . -print0 | cpio --verbose --null --create --format=newc | gzip -7 | tee "$target" >/dev/null 2>/dev/null
```
Comment (pzakha) [inline scripts/stage1.sh:76]: Why use `tee` if you are redirecting output to `/dev/null`?

File: scripts/recovery_sync [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+
+#
+# This script will sync host keys, authorized public keys, and networking
+# configuration from the running sytem to the recovery environment. It should
```
Comment (pzakha) [inline scripts/recovery_sync:5]: typo: sytem

File: scripts/recovery_sync [hunk 29]
```diff
@@ -0,0 +1,37 @@
...
+chmod 0755 .
+gunzip <"$target" | cpio -idmv
+
+rsync -a /run/systemd/network/ etc/systemd/network
+mkdir -p etc/dropbear
+for fmt in $(find /etc/ssh -name \*key | sed 's/.*host_\(.*\)_key/\1/'); do
+	# Unfortunately, dropbear doesn't support ed25519 keys.
+	[[ "$fmt" == "ed25519" ]] && continue
+	./bin/dropbearconvert openssh dropbear "/etc/ssh/ssh_host_${fmt}_key" "etc/dropbear/dropbear_${fmt}_host_key"
+done
```
Comment (pzakha) [inline scripts/recovery_sync:32]: What does this do?

File: scripts/recovery_sync [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+
+rsync -a /etc/machine-id etc/
+rsync -a /export/home/delphix/.ssh export/home/delphix/
+
+find . -print0 | cpio --verbose --null --create --format=newc | gzip -7 | tee "$target" >/dev/null
```
Comment (pzakha) [inline scripts/recovery_sync:37]: Why use `tee` here?

File: scripts/stage1.sh (hunk 26 above)
Comment (pcd1193182) [reply scripts/stage1.sh:26]: I've tweaked this so it's going to copy preferentially from the workdir. It's unfortunately not possible to force ldd to only search in the workdir, but I can make it always search the workdir paths first.

File: scripts/stage1.sh (hunk 27 above)
Comment (pcd1193182) [reply scripts/stage1.sh:80]: This was important when the script was using sudo, since you can't easily cause the redirect to act as a different user; `sudo tee foo >/dev/null` is an easy way to write the incoming pipe to foo as root without running all the other commands as root. Since the switchover, it's not necessary anymore; I've fixed it here and elsewhere.

File: scripts/recovery_sync (hunk 29 above)
Comment (pcd1193182) [reply scripts/recovery_sync]: This converts the system host keys to a format that dropbear can read. I'll add a comment.

File: scripts/delphix-bootcount.service [hunk 31]
```diff
@@ -0,0 +1,16 @@
...
+Requires=ssh.service
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStartPre=/bin/sleep 300
+ExecStart=/usr/bin/grub-editenv /boot/grub/grubenv set bootcount=0
+
+[Install]
+WantedBy=delphix.target
```
Comment (pzakha) [inline scripts/delphix-bootcount.service:16]: missing new line

File: scripts/delphix-bootcount.service [hunk 32]
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) 2020 by Delphix. All rights reserved.
+#
+
+[Unit]
+Descrption=Delphix bootcount reset service
+Requires=ssh.service
```
Comment (pzakha) [inline scripts/delphix-bootcount.service:7]: You should also add `After=ssh.service`, as it is not implicit.

File: scripts/42_recovery [hunk 33]
```diff
@@ -0,0 +1,25 @@
+#! /bin/bash
```
Comment (pzakha) [inline scripts/42_recovery:1]: Whitespace after #!
